### PR TITLE
[QeRL] Layerwise Reloading

### PR DESCRIPTION
**Pull request by kylesayrs**
## Purpose ##
* Support weight reloading of quantized models
  * Improves upon https://github.com/vllm-project/vllm/pull/28480
    * Changes are more maintainable
      * FP8 can now do arbitrary kernel processing (padding, renaming, deepgemm, ect.)
      * FP8 does not need to reattach weight loaders after processing
      * Can remove complications around renaming weight_scale_inv
    * Support reloading with FP8 marlin
  * Improves upon https://github.com/vllm-project/vllm/pull/23014
    * Reduce memory requirements by 3x for FP8
    * Expand support to all quant configs, not just torchao
  * Support reloading of (theoretically any quantization scheme)
    * INT4 has been tested
    * FP8_BLOCK has been tested
    * FP8_DYNAMIC has been tested
    * NVFP4_A16 has been tested

### Nomenclature ##
<details>

- “Checkpoint format” refers to the format in which weights are loaded from disk or provided by a user.
- “Model format” refers to the state of the model after `init` but before weights are processed with `process_weights_after_loading` . The mapping between “checkpoint format” and “model format” is implemented by `model.load_weights`.
- “Kernel format” refers to the state of the model after `process_weights_after_loading`
- In the case that checkpoint format is unquantized, but the kernel format is quantized, we call this “online quantization”, where unquantized weights are quantized by vLLM during/after loading.
</details>

## Design ##
1. Weight metadata (shape, attributes) are captured before kernel processing
2. Kernel processing occurs
3. User triggers reloading
4. Weight metadata is used to restore the layer to original load-ready model format
5. Loaded weights are cached
6. Once all loaded weights are ready for a given layer
  a. the layer is materialized on device
  b. the weights are loaded into the layer
  c. the layer is processed into kernel format
  d. the new tensors are copied into the original tensor storage
7. Any remaining layers which did not load all weights are cleaned up

## Changes ##
* Implement 4 public functions
  * `record_metadata_for_reloading` is called at the end of `initialize_model`
  * `initialize_layerwise_reload` is called before weights are loaded a second time
  * `finalize_layerwise_reload` is called after weights are loaded a second
  * `support_quantized_model_reload_from_hp_weights` is a decorator which calls `initialize_layerwise_reload` and `finalize_layerwise_reload`, preserves backwards compatibility for torchao users
* Track layer information (restore metadata, kernel tensors, loading progresss) via the global dictionary `LAYER_RELOADING_INFO`
  * Before and after weight reloading, this dictionary only holds meta tensors. This dictionary is a weak key dictionary, so those meta tensors are garbage collected if the model is garbage collected.
* Expand capabilities of `lm.collective_rpc("reload_weights")`
  * Can now pass `weights_iterator` directly
  * Alternatively, can pass `weights_path` to point to another checkpoint on disk
  * `checkpoint_format` controls whether the weights are provided in model format (default) or kernel format (quantized, renamed, and sharded)
* Misc
  * Improve `generate_prompt_perplexity` by allowing users to pass a mask

## Testing ##
* Add unit tests for materializing and restoring layers
* Add model tests for full precision, int4, and fp8 weight updating
* Add model test for MoE weight updating
* Modify `test_online_quant_config_dict_json` to actually test online quantization via the `model.load_weights` pathway

## Future Work ##
* Implement composing with the following features
  * Attention/MLA quantization
  * Expert-parallel load balancing (all that should be needed is to pause async EPLB during reload)
  * Reuse code and compose with https://github.com/vllm-project/vllm/pull/31914
  * Distributed weight transfer and updating: #31943
  * Remove logic from https://github.com/vllm-project/vllm/pull/28480

### CODE REVIEW

File: vllm/model_executor/model_loader/reload_utils.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            # materialize meta tensor if necessary
+            if param.device == torch.device("meta"):
+                param = materialize_meta_tensor(param)
+                setattr(layer, name, param)
+
+            # load weight into target
+            original_weight_loader(param, loaded_weight)
+
+            # process and copy when weights are loaded
+            load_numel_remaining -= loaded_weight.numel()
```
Comment (vkuzo) [inline vllm/model_executor/model_loader/reload_utils.py]: something like https://github.com/vllm-project/vllm/pull/30900 is needed to handle various edge cases when counting updated elements

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload_utils.py]: Totally right, thanks for pointing this out. I'll want to change the context a big to guarantee that we're only counting copies into the target weight (to avoid counting other, unrelated copies that weight loader might be doing)

File: vllm/model_executor/model_loader/reload_utils.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+import torch
+
+from vllm.logger import init_logger
+from vllm.model_executor.model_loader.utils import process_weights_after_loading
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+
+logger = init_logger(__name__)
+
+RESTORE_ATTRS = ["weight_loader"]
```
Comment (jerryzh168) [inline vllm/model_executor/model_loader/reload_utils.py]: seems `tensor.to("meta")` is dropping custom attributes, I think we need more attributes here like `input_dim` / `output_dim` etc. (https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/torchao.py#L337) for the weight loading to work?

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload_utils.py]: Thanks for the callout! Can you point me to where these attributes are read? They might not be needed for reloading, but if they are, we should include them here.

Comment (jerryzh168) [reply vllm/model_executor/model_loader/reload_utils.py]: yeah e.g. `output_dim` is used during weight loading i.e. `original_weight_loader(param, loaded_weight)` I think:

https://github.com/vllm-project/vllm/blob/69f8a0ea37a018b2e54897c44f8832de51a2bf5c/vllm/model_executor/layers/linear.py#L552

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload_utils.py]: There's a lot more than just these attributes. I decided to implement `to_meta_tensor` which copies the entire `__dict__` as well as the `__class__`.

We might have to replicate your method unbinding/rebinding logic if we find that there are dict methods, but I don't think this is a common case: All the attributes I've seen so far are not bound to the original tensor.

File: vllm/model_executor/model_loader/reload_utils.py (hunk 1 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload_utils.py]: I ended up copying your `CopyNumelCounter` implementation and tweaking it so that it can support meta tensors. That way, we can determine how many tensors will be loaded by a given `weight_loader` call, without the memory/performance cost of actually loading the weight.

File: vllm/model_executor/model_loader/reload/meta.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            kwargs = {}
+
+        if func is torch.ops.aten.copy_.default and args[0].device.type == "meta":
+            assert args[0].numel() == args[1].numel()
+            self.copied_numel += args[0].numel()
+
+        return func(*args, **kwargs)
+
+
+def get_numel_loaded(weight_loader: Callable, args: inspect.BoundArguments) -> int:
```
Comment (vkuzo) [inline vllm/model_executor/model_loader/reload/meta.py]: this is neat!

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+logger = init_logger(__name__)
+
+__all__ = [
+    "record_metadata_for_reloading",
+    "initialize_layerwise_reload",
+    "finalize_layerwise_reload",
+]
+
+
+LAYER_RELOADING_INFO: dict[torch.nn.Module, LayerReloadingInfo] = dict()
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py]: Will address by using a weak key dictionary

File: tests/conftest.py [hunk 5]
```diff
@@ -1034,13 +1036,21 @@ def generate_prompt_perplexity(self, prompts: list[str]) -> list[float]:
...
-        for output in outputs:
+        for prompt_index, output in enumerate(outputs):
             output = cast(TokensTextLogprobsPromptLogprobs, output)
             token_datas = cast(list[dict[int, Logprob] | None], output[3])
             assert token_datas[0] is None
+
             token_log_probs = []
-            for token_data in token_datas[1:]:
+            mask_prefix_len = mask_prefix_lens[prompt_index]
+            for token_data in token_datas[mask_prefix_len:][1:]:
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply tests/conftest.py]: Valid, will address soon

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            # Best solution is to ensure that `load_numel_total` reflects the
+            # actual number of weights loaded, either by modifying qconfigs to
+            # create as many weights as loaded (see padding issue as well)
+            # or maybe capturing how many weights are loaded on first pass
+            #
+            # For now, `load_numel_total` is still safe to use as long as
+            # there's no way to reach `load_numel_total` without loading all
+            # necessary weights. `weight_shape` is very small, so this is safe.
+            # see Limitations(4)
+            return
```
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/layerwise.py:137]: Should we add a debug log here if we run into issues, or would that even be too spammy?

File: vllm/v1/worker/gpu_model_runner.py [hunk 7]
```diff
@@ -4008,13 +4014,93 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
+        model = self.get_model()
+        weights_to_load = {name for name, _ in model.named_parameters()}
+        counter_before_reloading = time.perf_counter()
+
+        # load weights from disk if none are provided
+        if weights_iterator is None:
+            model_loader = get_model_loader(self.load_config)
+            if not hasattr(model_loader, "get_all_weights"):
+                raise NotImplementedError(
+                    f"Model reloading with `{self.load_config.load_format}` format"
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py:4251]: nit: should probably say "is not implemented" explicitly

File: vllm/v1/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -4008,13 +4014,93 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
+                )
+
+            if weights_path is not None:
+                self.model_config.model = weights_path
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py:4255]: Why is this needed? Also do we need to do validation on this path? It seems like a late state to change this shared config

File: vllm/v1/worker/gpu_model_runner.py [hunk 9]
```diff
@@ -4008,13 +4014,93 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
         return None
 
-    def reload_weights(self) -> None:
-        assert getattr(self, "model", None) is not None, (
-            "Cannot reload weights before model is loaded."
+    def reload_weights(
+        self,
+        weights_iterator: Iterable[tuple[str, torch.Tensor]] | None = None,
+        weights_path: str | None = None,
+        checkpoint_format: bool = True,
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py]: nit: `is_checkpoint_format` since bool

File: vllm/model_executor/model_loader/reload/types.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+class LayerReloadingInfo:
+    # model format (meta), populated by `record_metadata_for_reloading`
+    restore_metadata: LayerTensors = field(default_factory=lambda: ({}, {}))
+
+    # kernel format (device)
+    kernel_tensors: LayerTensors = field(default_factory=lambda: ({}, {}))
+
+    # track how many restored elements are ready for loading
+    load_numel: int | float = 0
+    load_numel_total: int | float = float("inf")
```
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/types.py]: Why are these floats? I'd rather just use `int | None`

File: vllm/model_executor/model_loader/reload/__init__.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+Limitations:
+1. Composition with CPU offloading has not been implemented
+2. Reloading Attention/MLA weights (q_scale, k_scale, v_scale) has not been implemented
+3. Tied parameters will only reflect processing from one of the parent layers (for
+   example, only processing from embed_tokens will have an effect)
+4. This design assumes that the number of weights loaded from disk is the same as the
+   number of weights created at model init time. This is not true for quant methods
+   which (1) pad weights or (2) load qkv weights into the same parameter. Both of these
+   cases are non-issues for today's quant methods, but future quantizations may cause
+   reloading to fail
```
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/__init__.py:18]: We should open a tracking issue for some limitations, especially if planned for future work

File: vllm/model_executor/model_loader/reload/utils.py
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/utils.py:1]: nit: we generally use `utils.py`

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+    "get_layerwise_info",
+    "record_metadata_for_reloading",
+    "initialize_layerwise_reload",
+    "finalize_layerwise_reload",
+]
+
+
+LAYERWISE_INFO: WeakKeyDictionary[torch.nn.Module, LayerReloadingInfo] = (
+    WeakKeyDictionary()
+)
```
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/layerwise.py:42]: Worth a comment on the structure of this global dict

File: vllm/v1/worker/gpu_model_runner.py [hunk 13]
```diff
@@ -4008,13 +4014,93 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
+        with torch.device(load_device):
+            if checkpoint_format:
+                # @kylesayrs @jerryzh168: disable torchao reloading
+                # assume `reload_weights` is the target entrypoint
+                model._do_torchao_reload = False
+
+                # load weights from checkpoint/ original model format
+                initialize_layerwise_reload(model)
+                loaded_weights = model.load_weights(weights_iterator)
+                finalize_layerwise_reload(model, self.model_config)
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py:4271]: This assumes weights are loaded with a single iter pass since `initialize_layerwise_reload` wraps all loaders and expects all weights. If https://github.com/vllm-project/vllm/pull/31943 calls update_weights multiple times, the load_numel tracking in might trigger layerwise_process mid-way through loading, before all weights are ready. What is the plan for compat there?

File: vllm/model_executor/model_loader/reload/layerwise.py (hunk 6 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py:137]: This code is expected to run for every marlin layer. I'll add a debug message that triggers only once.

File: vllm/model_executor/model_loader/reload/torchao_decorator.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        **kwargs,
+    ):
+        model = self.module
+
+        if not getattr(model, "_do_torchao_reload", False):
+            return original_load_weights(self, weights, *args, **kwargs)
+
+        initialize_layerwise_reload(model)
+        loaded_weights = original_load_weights(self, weights, *args, **kwargs)
+        finalize_layerwise_reload(model, model._model_config)
```
Comment (mgoin) [inline vllm/model_executor/model_loader/reload/torchao_decorator.py:54]: Why do we still need this decorator for torchao and can't just call the layer functions directly?

File: vllm/v1/worker/gpu_model_runner.py [hunk 15]
```diff
@@ -4008,13 +4014,93 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
+        # begin loading weights
+        logger.info_once("Reloading weights inplace...", scope="local")
+        load_device = (
+            self.vllm_config.load_config.device or self.vllm_config.device_config.device
+        )
+        with torch.device(load_device):
+            if checkpoint_format:
+                # @kylesayrs @jerryzh168: disable torchao reloading
+                # assume `reload_weights` is the target entrypoint
+                model._do_torchao_reload = False
```
Comment (mgoin) [inline vllm/v1/worker/gpu_model_runner.py]: Can this be handled by initialize_layerwise_reload?

File: vllm/v1/worker/gpu_model_runner.py (hunk 8 above)
Comment (kylesayrs) [reply vllm/v1/worker/gpu_model_runner.py:4255]: The ability to pass `weights_path` is a nice-to-have feature, makes testing easier and allows users to pass directory paths, which is a real use case.

Do you think we need to change anything more than `self.model_config.model`?

File: vllm/model_executor/model_loader/reload/types.py (hunk 10 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/types.py]: Probably safe to make this change now.

File: vllm/v1/worker/gpu_model_runner.py (hunk 13 above)
Comment (kylesayrs) [reply vllm/v1/worker/gpu_model_runner.py:4271]: Yes, your characterization is correct and required by any layerwise design.

The [integration plan](https://github.com/vllm-project/vllm/pull/31943#pullrequestreview-3667541304) is as follows:
`/init_weight_transfer` -> `initialize_layerwise_reload(model)`
...
`/update_weights` -> `model.load_weights(batch_one)`
`/update_weights` -> `model.load_weights(batch_two)`
...
`/finalize_weight_update` -> `finalize_layerwise_reload(model, self.model_config)`

File: vllm/model_executor/model_loader/reload/torchao_decorator.py (hunk 14 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/torchao_decorator.py:54]: Not sure if I understand the question. Torchao's online FP8 implementation requires restoring logic. Without this decorator, users who use `model.load_weights` to do online torchao quantization would have their flows break.

File: vllm/v1/worker/gpu_model_runner.py (hunk 15 above)
Comment (kylesayrs) [reply vllm/v1/worker/gpu_model_runner.py]: Good call.

File: vllm/model_executor/model_loader/reload/__init__.py (hunk 11 above)
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/__init__.py:18]: https://github.com/vllm-project/vllm/issues/33038

**Review by mgoin — APPROVED** [review]
Awesome work, LGTM!
